File: src/main/scala/category/stack/_735_Solution.scala
package category.stack

import scala.collection.mutable

/**
  * 735. Asteroid Collision
  *
  * [5,10,-5]
  * [-2,1]
  * [8,-8]
  * [-8,8]
  * [10,2,-5]
  * [-2, -1, 1, 2]
  * []
  */
object _735_Solution {
  def asteroidCollision(asteroids: Array[Int]): Array[Int] = {
    val stack = new mutable.Stack[Int]
    asteroids.foreach { item =>
      var curr = item
      while (stack.nonEmpty && curr < 0 && stack.top > 0) {
        if (Math.abs(curr) < Math.abs(stack.top)) {curr = 0}
        else if (Math.abs(curr) == Math.abs(stack.top)) {stack.pop; curr = 0}
        else {stack.pop; curr = item}
      }
      if (curr != 0) stack.push(item)
    }
    stack.toArray.reverse
  }
}

File: src/main/python/medium/_700_750/_735_Solution.py
# 735. Asteroid Collision
from typing import List


class _735_Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        if not asteroids: return asteroids
        stack = []
        for asteroid in asteroids:
            while stack and asteroid < 0 < stack[-1]:
                if abs(asteroid) <= abs(stack[-1]):
                    if abs(stack[-1]) == abs(asteroid): stack.pop()
                    break
                else:
                    stack.pop()
                    continue
            else:
                stack.append(asteroid)
        return stack


if __name__ == '__main__':
    tar = _735_Solution()
    assert tar.asteroidCollision([5, 10, -11])
    assert tar.asteroidCollision([5, 10, -5])
    assert tar.asteroidCollision([-2, -1, 1, 2])
